src/plugins/action-queue/types.js

```diff
@@ -0,0 +1,171 @@
+// @flow
+/* eslint-disable no-use-before-define */
+
+//
+// Action Operations
+//
+
+export type ActionOp =
+  | {
+      type: 'seq',
+      actions: ActionOp[]
+    }
+  | {
+      type: 'par',
+      actions: ActionOp[]
+    }
+  | {
+      type: 'exchange:buy',
+      nativeAmount: string,
+      walletId: string,
+      tokenId?: string,
+      exchangePluginId: string
+    }
+  | {
+      type: 'exchange:sell',
+      nativeAmount: string,
+      walletId: string,
+      tokenId?: string,
+      exchangePluginId: string
+    }
+  | {
+      type: 'loan:borrow',
+      borrowPluginId: string,
+      nativeAmount: string,
+      walletId: string,
+      tokenId?: string
+    }
+  | {
+      type: 'loan:deposit',
+      borrowPluginId: string,
+      nativeAmount: string,
+      walletId: string,
+      tokenId?: string
+    }
+  | {
+      type: 'loan:repay',
+      borrowPluginId: string,
+      nativeAmount: string,
+      walletId: string,
+      tokenId?: string
+    }
+  | {
+      type: 'loan:withdraw',
+      borrowPluginId: string,
+      nativeAmount: string,
+      walletId: string,
+      tokenId?: string
+    }
+  | {
+      type: 'swap',
+      fromWalletId: string,
+      toWalletId: string,
+      fromTokenId?: string,
+      toTokenId?: string,
+      nativeAmount: string
+    }
+  // Useful for development/testing
+  | {
+      type: 'toast',
+      message: string
+    }
+  | {
+      type: 'delay',
+      ms: number
+    }
+
+//
+// Action Effects
+//
+
+export type ActionEffect =
+  | {
+      type: 'seq',
+      opIndex: number,
+      childEffect: ActionEffect
+    }
+  | {
+      type: 'par',
+      childEffects: ActionEffect[]
+    }
+  | {
+      type: 'balance',
+      address?: string, // Address balance if present otherwise wallet balances
+      aboveAmount?: string,
+      belowAmount?: string,
+      walletId: string,
+      tokenId?: string
+    }
+  | {
+      type: 'tx-confs',
+      txId: string,
+      walletId: string,
+      confirmations: number
+    }
+  | {
+      type: 'price',
+      currencyPair: string,
+      aboveRate?: string,
+      belowRate?: string
+    }
+  | {
+      type: 'end',
```

Comment:
The name "end" implies "end of code", which is not what we mean. We really mean "done" or "finished" or a word more like that.

Reply:
I can live with `"done"`.